### PR TITLE
Add bubble and insertion sort algorithms

### LINKED ISSUE #417: Lab 02 - Bubble Sort
# Lab 2 - Bubble sort

- Create an executable go program in directory `02_bubble/USERNAME`
- Write a function that returns a sorted copy of `int` slice `s` using [bubble sort](https://en.wikipedia.org/wiki/Bubble_sort):

```
func bubble(s []int) []int
```

Call `fmt.Println(bubble([]int{3, 2, 1, 5}))` in `main` to print:

```
[1 2 3 5]
```

- Bonus points: implement [Insertion sort](https://en.wikipedia.org/wiki/Insertion_sort)
- Extra bonus points: implement an _O(n_ _log(n))_ sorting algorithm

### PR DESCRIPTION
**Pull request by patrickmarabeas**
Implements the bubble and insertion sort algorithms. Both functions take a slice of n ints and return a sorted slice of n ints using specified sorting algorithm.

- Adds bubble sort algorithm
- Adds insertion sort algorithm

Fixes #417 

Review of colleague's PR #501

### CODE REVIEW

**Comment by Marchitect** [comment]
Looks all good to me, I think your solutions look neat, good work. One suggestion I can give would be that you might consider adding negative integers into test cases.

**Comment by patrickmarabeas** [comment]
@Marchitect - Added negative integer tests.

**Review by undrewb — DISMISSED** [review]
Nice neat implementation. Plenty of test cases. Good work!

File: 02_bubble/patrickmarabeas/main.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+)
+
+var out io.Writer = os.Stdout
+
+// Bubble implements the Bubble sort algorithm
+func bubble(s []int) []int {
+	length := len(s)
+	// Create copy
+	result := make([]int, length)
+	copy(result, s)
```
Comment (anzdaddy) [inline 02_bubble/patrickmarabeas/main.go:16]: This forces a copy that the caller may not require. Making a copy should be left up to the caller.

Comment (patrickmarabeas) [reply 02_bubble/patrickmarabeas/main.go:16]: This is a good point, thanks. I'll make sure I'm not being opinionated re immutability in the future.

File: 02_bubble/patrickmarabeas/main.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+// Bubble implements the Bubble sort algorithm.
+func bubble(s []int) []int {
```
Comment (anzdaddy) [inline 02_bubble/patrickmarabeas/main.go:12]: nit: With the change in mutability behaviour, it would be clearer to not return anything.

**Review by anzdaddy — DISMISSED** [review]

File: 02_bubble/patrickmarabeas/main.go (hunk 2 above)
Comment (patrickmarabeas) [reply 02_bubble/patrickmarabeas/main.go:12]: ~Would this result in the following signature `func bubble(s *[]int)`?~

I see Slices are passed by reference already.

Comment (patrickmarabeas) [reply 02_bubble/patrickmarabeas/main.go:12]: Added alternative implementation and tests

File: 02_bubble/patrickmarabeas/main_test.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+				t.Errorf("\nExpected: %d\nGot:      %d", expected, got)
+			}
+		})
+	}
+}
+
+func TestBubble2(t *testing.T) {
+	for name, c := range cases {
+		got := c.input
+		bubble2(got)
```
Comment (anzdaddy) [inline 02_bubble/patrickmarabeas/main_test.go]: Don't forget that you're now mutating the original and thus might require a copy.

**Review by juliaogris — COMMENTED** [review]
Please address [failing build]().  
From the [lab assignment doc](https://github.com/anz-bank/go-course/blob/master/02_bubble/README.md):
> Write a function that returns a sorted *copy* 
(my emphasis).  

You are mutating the input.

File: 02_bubble/patrickmarabeas/main_test.go (hunk 3 above)
Comment (patrickmarabeas) [reply 02_bubble/patrickmarabeas/main_test.go]: Reverted to returning a copy, as per spec

**Comment by patrickmarabeas** [comment]
Reverted to original implementation

**Review by juliaogris — APPROVED** [review]
LGTM.

Just a minor nit on style, don't comment the obvious, that just creates noise:

```go
	// Create copy
	result := make([]int, length)
	copy(result, s)
```

**Pull request merged**